grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java

```diff
@@ -0,0 +1,323 @@
+
+  // Server list states
+  @GuardedBy("lock")
+  private HashMap<SocketAddress, ResolvedServerInfo> servers;
+  @GuardedBy("lock")
+  @VisibleForTesting
+  private RoundRobinServerList roundRobinServerList;
+
+  private ExecutorService executor;
+  private ScheduledExecutorService deadlineCancellationExecutor;
+
+  GrpclbLoadBalancer(String serviceName, TransportManager tm) {
+    this.serviceName = serviceName;
+    this.tm = tm;
+    executor = SharedResourceHolder.get(GrpcUtil.SHARED_CHANNEL_EXECUTOR);
+    deadlineCancellationExecutor = SharedResourceHolder.get(GrpcUtil.TIMER_SERVICE);
+  }
+
+  @VisibleForTesting
+  StreamObserver<LoadBalanceResponse> getLbResponseObserver() {
+    synchronized (lock) {
+      return lbResponseObserver;
+    }
+  }
+
+  @VisibleForTesting
+  RoundRobinServerList getRoundRobinServerList() {
+    synchronized (lock) {
+      return roundRobinServerList;
+    }
+  }
+
+  @Override
+  public ListenableFuture<ClientTransport> pickTransport(@Nullable RequestKey requestKey) {
+    RoundRobinServerList serverListCopy;
+    synchronized (lock) {
+      Preconditions.checkState(!closed, "already closed");
+      if (roundRobinServerList == null) {
+        if (lastError == null) {
+          return pendingPicks.newBlankFuture();
+        } else {
+          return Futures.immediateFailedFuture(lastError);
+        }
+      }
+      serverListCopy = roundRobinServerList;
+    }
+    return serverListCopy.getTransportForNextServer();
+  }
+
+  @Override
+  public void handleResolvedAddresses(
+      List<ResolvedServerInfo> updatedServers, Attributes config) {
+    synchronized (lock) {
+      ArrayList<SocketAddress> addrs = new ArrayList<SocketAddress>(updatedServers.size());
+      for (ResolvedServerInfo serverInfo : updatedServers) {
+        addrs.add(serverInfo.getAddress());
+      }
+      EquivalentAddressGroup newLbAddresses = new EquivalentAddressGroup(addrs);
+      if (!newLbAddresses.equals(lbAddresses)) {
+        lbAddresses = newLbAddresses;
+        connectToLb();
+      }
+    }
+  }
+
+  @GuardedBy("lock")
+  private void connectToLb() {
+    if (closed) {
+      return;
+    }
+    lbResponseObserver = null;
+    // TODO(zhangkun83): should use a separate authority for LB servers
+    Preconditions.checkNotNull(lbAddresses, "lbAddresses");
+    ListenableFuture<ClientTransport> transportFuture = tm.getTransport(lbAddresses);
+    Futures.addCallback(
+        Preconditions.checkNotNull(transportFuture),
+        new FutureCallback<ClientTransport>() {
+          @Override public void onSuccess(ClientTransport transport) {
+            synchronized (lock) {
+              if (closed) {
+                return;
+              }
+              lbTransport = transport;
+              startNegotiation();
+            }
+          }
+
+          @Override public void onFailure(Throwable t) {
+            handleError(Status.fromThrowable(t).augmentDescription(
```

Comment:
There is no recovery for this error. It looks like this will fail all future RPCs unless the addresses change such that `handleResolvedAddresses()` is called again.